core/node/api_server/src/tx_sender/gas_estimation.rs

```diff
@@ -0,0 +1,511 @@
+            {
+                Some(balance) => balance,
+                None => self.get_balance(&tx.initiator_account()).await?,
+            };
+
+            if tx.execute.value > balance {
+                tracing::info!(
+                    "fee estimation failed on validation step.
+                    account: {} does not have enough funds for for transferring tx.value: {}.",
+                    tx.initiator_account(),
+                    tx.execute.value
+                );
+                return Err(SubmitTxError::InsufficientFundsForTransfer);
+            }
+        }
+        Ok(())
+    }
+
+    fn adjust_search_bounds(
+        lower_bound: &mut u64,
+        upper_bound: &mut u64,
+        pivot: u64,
+        result: &VmExecutionResultAndLogs,
+    ) {
+        // For now, we don't discern between "out of gas" and other failure reasons since it's difficult in the general case.
+        if result.result.is_failed() {
+            *lower_bound = pivot + 1;
+        } else {
+            *upper_bound = pivot;
+        }
+    }
+}
+
+/// Initial gas estimate with effectively infinite gas limit.
+#[derive(Debug)]
+pub(super) struct InitialGasEstimate {
+    /// Set to `None` if not estimated (e.g., for L1 transactions), or if the VM returned bogus refund stats.
+    pub total_gas_charged: Option<u64>,
+    /// Set to `None` if not estimated (e.g., for L1 transactions).
+    pub computational_gas_used: Option<u64>,
+    pub operator_overhead: u64,
+    pub gas_charged_for_pubdata: u64,
+}
+
+impl InitialGasEstimate {
+    /// Total gas charged for a transaction consists of:
+    ///
+    /// - Operator-set overhead (`operator_overhead`)
+    /// - Intrinsic bootloader overhead
+    /// - Gas used during validation / execution
+    /// - Gas charged for pubdata at the end of execution (`gas_for_pubdata`)
+    ///
+    /// We add `operator_overhead` manually to the binary search argument at each `step()` because it depends on the gas limit in the general case,
+    /// so this value corresponds to the other 3 terms.
+    ///
+    /// If the value cannot be computed, it is set to `None`.
+    pub fn lower_gas_bound_without_overhead(&self) -> Option<u64> {
+        let mut total_gas_bound = self.computational_gas_used? + self.gas_charged_for_pubdata;
+        if let Some(gas_charged) = self.total_gas_charged {
+            total_gas_bound = total_gas_bound.min(gas_charged);
+        }
+        total_gas_bound.checked_sub(self.operator_overhead)
+    }
+
+    pub fn optimistic_gas_limit_without_overhead(&self) -> Option<u64> {
+        let gas_charged_without_overhead = self
+            .total_gas_charged?
+            .checked_sub(self.operator_overhead)?;
+        // 21/20 is an empirical multiplier. It is higher than what empirically suffices for some common transactions;
+        // one can argue that using 64/63 multiplier would be more accurate due to the 63/64 rule for far calls
```

Comment:
More precisely, judging by unit tests, the only exception I've found so far is a transaction decommitting a very large bytecode (~320kB).